Caltech-IPAC/firefly
# Dm 10855 UI for accessing user workspace

This development includes, 
- Implementing three FilePicker components to
   . listing the files in workspace, read files from workspace, save files to workspace. 
   . management of the files (delete file/rename file/add subfolder for generic FilePicker)
- implementing components to selecting file upload/download from either local file or workspace file system. 
-  clone and modify source codes from react-keyed-file-browser. https://github.com/ABASystems/react-keyed-file-browser, which is used as the base for the file picker. The set of source codes is placed under src/firefly/js/externalSource/FilePicker/react-keyed-file-browser/. The modification is based on the needs of workspace file access. 
- modify workspace management and firefly codes at the server side to integrate UI development to access workspace files including fixing the request parameters, server command processing, and the authentication issues. 
- Update the UI upload and download codes across Firefly implementation to accommodate the new implementation  for workspace file access. 


test: 
- package.json is changed, please update node_modules. 
- there is no login in Firefly, please login to irsa.ipac.caltech.edu by using username test@ipac.caltech.edu.
- start Firefly by going to <your host name>.ipac.caltech.edu:8080/firefly

test file management on workspace files, 
- select tab 'Workspace' (this tab is mainly for the demo of how to use FilePicker component) 
   click 'Show workspace below' => a list of files under https://irsa.ipac.caltech.edu/ssospace/test@ipac.caltech.edu/ is shown. 
- try file management (add folder, delete/rename file) on the FilePicker by selecting the item and the action on the action bar. 

test 'save file to workspace' 
- do image search or catalog search by selecting tab 'Images' or 'Catalogs', click the disk-like icon to save the image or table to workspace. 
- from the download popup, select file location between 'Local file' and 'Workspace' and enter the file name to 'save as'.  If 'Workspace' is selected, a FilePicker is rendered below for choosing the folder to save the table/image to. 

test 'upload file from workspace' (no file analysis report)
- From Firefly viewer, go to tab 'Images' and then 'FITS File' or tab 'Catalogs' and then 'Load Catalog' to get the image file or table file from either 'Local File' or 'Workspace'. Click 'Workspace Upload' button if radio 'Workspace' is selected to get a 'Read file from Workspace' popup. 
- select the file from the FilePicker of the popup, and click 'open', then the selected workspace file path will be shown by the 'Workspace upload' button. 

test 'Upload file from workspace' (with file analysis report) 
- go to tab 'Upload' 
- select 'Upload from workspace', click 'Workspace upload' button to get FilePicker popup to select any file from workspace, click 'Open' or 'cancel' to exit the popup. 
- the file analysis report is shown under as what 'Upload File' and 'Upload from URL' provide.
